Finish the rest of the event documentation

MCJack123 opened the pull request:
This PR adds the rest of the events to the docs. The data was fetched mostly from the old wiki, with new changes added from the new wiki, as well as some things I've remembered over time.

Sorry for the large amount of commits; I was too lazy to clone it myself, and just did everything in the GitHub web interface. Just squash it (you usually do that anyway).

Closes: #681, closes #640

Lupus590 commented:
closes: #681
closes: #640

SquidDev commented on doc/events/alarm.md:
````diff
@@ -0,0 +1,17 @@
+```lua
+os.setAlarm(os.time() + 0.05)
+local event, id = os.pullEvent("alarm")
+print("Alarm with ID " .. id .. " was fired")
````
Is it worth doing this in a `while` loop and checking the IDs match? 

Obviously no other `alarm` events are going to be queued when running this, but just to demonstrate "good practice" or something.

SquidDev commented on doc/events/http_check.md:
````diff
@@ -0,0 +1,28 @@
+## Example
+Checks a valid URL:
+```lua
+http.checkURLAsync("http://www.example.com")
````
Writing examples for these is hard, because `checkURLAsync` has a somewhat ugly interface.

More precisely, ` http.checkURLAsync("ht:/")` _won't_ queue a `http_check` event, as it's a syntactically incorrect URL - it just returns `false, error`. `http_check` events are only queued when the allowed/denied check complete (it's asynchronous as it requires a DNS lookup).

I'd honestly just leave these examples off for now. It's such a useless event (who bulk checks URLs?!) and any actual example is just going to end up looking like the source for `http.checkURL`.

SquidDev commented on doc/events/http_failure.md:
````diff
@@ -0,0 +1,30 @@
+Prints an error why the website cannot be contacted:
+```lua
+http.request("http://this.website.does.not.exist")
+local event, url, err = os.pullEvent("http_failure")
````
As with `alarm`, these should definitely be checking the URL.

_Maybe_ worth using a `tweaked.cc`/`computercraft.cc` domain instead of httpbin? On the basis that I can probably commit to those being maintained for as long as this mod is :). Though I guess httpbin isn't probably going to vanish any time soon.

SquidDev commented on doc/events/http_success.md:
````diff
@@ -0,0 +1,22 @@
+Prints the content of a website (this may fail if the request fails):
+```lua
+http.request("http://www.example.com")
+local event, url, handle = os.pullEvent("http_success")
````
As with `alarm`, definitely worth checking the URL.

SquidDev commented on doc/events/modem_message.md:
````diff
@@ -0,0 +1,21 @@
+```lua
+while true do
+  local event, side, channel, replyChannel, message, distance = os.pullEvent("modem_message")
+  print("Message received on side " .. side .. " on channel " .. channel .. " (reply to " .. replyChannel .. ") from " .. distance .. " blocks away with message " .. tostring(message))
````
Maybe at this point it might be worth using a format string :p. It's a pretty unreadable line otherwise.

SquidDev commented on doc/events/task_complete.md:
````diff
@@ -0,0 +1,26 @@
+## Example
+Prints the results of an asynchronous command:
+```lua
+commands.execAsync("say Hello")
````
As with `alarm`, worth checking the ID here.

SquidDev commented on doc/events/task_complete.md:
```diff
@@ -0,0 +1,26 @@
+see: commands.execAsync To run a command which fires a task_complete event.
+---
+
+The @{task_complete} event is fired when a command run with @{commands.execAsync} or any function in @{commands.async} finishes.
```
Hrmrmr.

_Technically_, this is fired on any asynchronous peripheral call (e.g. `drive.setDiskLabel`). However, most of the time it's invisible, as the peripheral call waits for the event.

I don't know if this behaviour is worth documenting or not. Definitely should drop the "This event is normally only available on Command Computers." statement as that's wrong, but the rest of the time it's really just an implementation detail.

SquidDev commented on doc/events/timer.md:
````diff
@@ -0,0 +1,17 @@
+## Example
+Starts a timer and then prints its ID:
+```lua
+os.startTimer(2)
````
As with `alarm`, worth checking the ID.

SquidDev commented on doc/events/turtle_inventory.md:
```diff
@@ -0,0 +1,14 @@
+module: [kind=event] turtle_inventory
+---
+
+The @{turtle_inventory} event is fired when a Turtle's inventory is changed.
```
```suggestion
The @{turtle_inventory} event is fired when a turtle's inventory is changed.
```

SquidDev commented on doc/events/terminate.md:
```diff
@@ -0,0 +1,25 @@
+module: [kind=event] terminate
+---
+
+The @{terminate} event is fired when Ctrl-T is held down.
```
```suggestion
The @{terminate} event is fired when <kbd>Ctrl-T</kbd> is held down.
```

SquidDev commented on doc/events/websocket_closed.md:
````diff
@@ -0,0 +1,16 @@
+Prints a message when a WebSocket is closed (this may take a minute):
+```lua
+local ws = http.websocket("ws://echo.websocket.org")
+local event, url = os.pullEvent("websocket_closed")
````
Worth checking the URL.

SquidDev commented on doc/events/websocket_failure.md:
````diff
@@ -0,0 +1,20 @@
+Prints an error why the website cannot be contacted:
+```lua
+http.websocketAsync("ws://this.website.does.not.exist")
+local event, url, err = os.pullEvent("websocket_failure")
````
Worth checking the URL.

SquidDev commented on doc/events/websocket_message.md:
````diff
@@ -0,0 +1,21 @@
+```lua
+local ws = http.websocket("ws://echo.websocket.org")
+ws.send("Hello!")
+local event, url, message = os.pullEvent("websocket_message")
````
Worth checking the URL.

SquidDev commented on doc/events/websocket_success.md:
````diff
@@ -0,0 +1,23 @@
+Prints the content of a website (this may fail if the request fails):
+```lua
+http.websocketAsync("ws://echo.websocket.org")
+local event, url, handle = os.pullEvent("websocket_success")
````
Worth checking the URL.

SquidDev reviewed (changes requested):
Thank you so much for doing this, I know how dull docs are.

I've left a load of review comments, but do just say if you want me to pick them up. I know I'm being fussy.

In 9ae0f4a993693ffc48ec8a4b5824c50f54026e0a, I/@BradyFromDiscord documented events' return values to include the event name.. This PR starts from the first "interesting" value. What are people's thoughts on this - is one clearer than the other? I have no preference, so would very much like someone else to have an opinion.

Lupus590 commented:
> In 9ae0f4a, I/@BradyFromDiscord documented events' return values to include the event name.. This PR starts from the first "interesting" value. What are people's thoughts on this?

Having the event name is my preference, it reflects what gets returned by `os.pullEvent` and means that new people might miss this trip up. `local id, message = is.pullEven("rednet_message")`

That a bit of a long winded way to say that listing the event name as the first return value reenphasises that all events have their name returned first from `os.pullEvent`

MCJack123 commented:
I've made the suggested changes. I've also implemented some of the fixes from #681.

> In 9ae0f4a, I/@BradyFromDiscord documented events' return values to include the event name.. This PR starts from the first "interesting" value. What are people's thoughts on this - is one clearer than the other? I have no preference, so would very much like someone else to have an opinion.

I typically tend to ignore the event name if filtering for a certain event, since it's somewhat redundant. However, I do agree with @Lupus590 that declaring the name with return values may help newcomers. I've gone ahead and added the event names to the return values.

However, some events don't have any parameters, and when writing them I didn't add any return values - I did not add the event name as a return value here, as there's no information that needs to be extracted from them (unless you are picking up multiple events, in which case `os.pullEvent` already gives you the event name). I'd like some feedback on whether that's a good idea, though.

SquidDev commented:
Thank you very much!

> However, some events don't have any parameters, and when writing them I didn't add any return values - I did not add the event name as a return value here. [...] I'd like some feedback on whether that's a good idea, though.

Yeah, I think that's reasonable enough.

BradyFromDiscord commented:
> In 9ae0f4a, I/@BradyFromDiscord documented events' return values to include the event name.. This PR starts from the first "interesting" value. What are people's thoughts on this - is one clearer than the other? I have no preference, so would very much like someone else to have an opinion.
I preferred stating the event name as the first return value as it was something that tripped me up when I first started using events. I think making it clear that the first return value is the event name itself is pertinent especially to newer players, but I don't mind how it's conveyed; if saying "every event's first return value is the name of said event, followed by the data attached to that event" is cleaner and clearer than being as verbose as my documentation is, then I'm perfectly fine with it ^^
